instance/tests/models/test_openedx_instance.py

```diff
@@ -0,0 +1,243 @@
+        self.assertEqual(instance.studio_domain, 'studio.sample.example.org')
+        self.assertEqual(instance.studio_url, 'http://studio.sample.example.org/')
+        self.assertEqual(str(instance), 'Sample Instance (sample.example.org)')
+
+    @patch_services
+    @patch('instance.models.openedx_instance.OpenEdXAppServer.provision', return_value=True)
+    def test_spawn_appserver(self, mocks, mock_provision):
+        """
+        Run spawn_appserver() sequence
+        """
+        instance = OpenEdXInstanceFactory(sub_domain='test.spawn', use_ephemeral_databases=True)
+
+        appserver_id = instance.spawn_appserver()
+        self.assertEqual(mock_provision.call_count, 1)
+
+        self.assertIsNotNone(appserver_id)
+        self.assertEqual(instance.appserver_set.count(), 1)
+        self.assertIsNone(instance.active_appserver)
+
+        # We are using ephemeral databases:
+        self.assertEqual(mocks.mock_provision_mysql.call_count, 0)
+        self.assertEqual(mocks.mock_provision_mongo.call_count, 0)
+        self.assertEqual(mocks.mock_provision_swift.call_count, 0)
+        # And DNS should never be changed from spawn_appserver() alone:
+        self.assertEqual(mocks.mock_set_dns_record.call_count, 0)
+
+        appserver = instance.appserver_set.get(pk=appserver_id)
+        self.assertEqual(appserver.name, "AppServer 1")
+        self.assertEqual(appserver.instance, instance)
+        for field_name in OpenEdXAppConfiguration.get_config_fields():
+            self.assertEqual(
+                getattr(instance, field_name),
+                getattr(appserver, field_name),
+            )
+        self.assertEqual(appserver.configuration_database_settings, "")
+        self.assertEqual(appserver.configuration_storage_settings, "")
+        configuration_vars = yaml.load(appserver.configuration_settings)
+        self.assertEqual(configuration_vars['COMMON_HOSTNAME'], instance.domain)
+        self.assertEqual(configuration_vars['EDXAPP_PLATFORM_NAME'], instance.name)
+        self.assertEqual(configuration_vars['EDXAPP_CONTACT_EMAIL'], instance.email)
+
+    @patch_services
+    def test_set_appserver_active(self, mocks):
+        """
+        Test set_appserver_active()
+        """
+        instance = OpenEdXInstanceFactory(sub_domain='test.activate', use_ephemeral_databases=True)
+        appserver_id = instance.spawn_appserver()
+        instance.set_appserver_active(appserver_id)
+        self.assertEqual(instance.active_appserver.pk, appserver_id)
+        self.assertEqual(mocks.mock_set_dns_record.mock_calls, [
+            call(name='test.activate', type='A', value='1.1.1.1'),
+            call(name='studio.test.activate', type='CNAME', value='test.activate'),
+        ])
+
+    @patch_services
+    @patch('instance.models.openedx_instance.OpenEdXAppServer.provision', return_value=True)
+    def test_spawn_appserver_names(self, mocks, mock_provision):
+        """
+        Run spawn_appserver() sequence
```

Comment:
@bradenmacdonald The docstring for this method is the same as the one for `test_spawn_appserver` above. Change to "Run spawn_appserver() sequence multiple times and check names of resulting app servers" perhaps?